go/vt/orchestrator/logic/topology_recovery.go

```diff
@@ -815,113 +613,120 @@ func checkAndRecoverDeadPrimary(analysisEntry inst.ReplicationAnalysis, candidat
 	if !(forceInstanceRecovery || analysisEntry.ClusterDetails.HasAutomatedPrimaryRecovery) {
 		return false, nil, nil
 	}
+	tablet, err := TabletRefresh(analysisEntry.AnalyzedInstanceKey)
+	if err != nil {
+		return false, nil, err
+	}
+
+	var candidateTabletAlias *topodatapb.TabletAlias
+	if candidateInstanceKey != nil {
+		candidateTablet, err := inst.ReadTablet(*candidateInstanceKey)
+		if err != nil {
+			return false, nil, err
+		}
+		candidateTabletAlias = candidateTablet.Alias
+	}
 	topologyRecovery, err = AttemptRecoveryRegistration(&analysisEntry, !forceInstanceRecovery, !forceInstanceRecovery)
 	if topologyRecovery == nil {
 		AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("found an active or recent recovery on %+v. Will not issue another RecoverDeadPrimary.", analysisEntry.AnalyzedInstanceKey))
 		return false, nil, err
 	}
 	log.Infof("Analysis: %v, deadprimary %+v", analysisEntry.Analysis, analysisEntry.AnalyzedInstanceKey)
 
-	// That's it! We must do recovery!
-	// TODO(sougou): This function gets called by GracefulPrimaryTakeover which may
-	// need to obtain shard lock before getting here.
-	unlock, err := LockShard(analysisEntry.AnalyzedInstanceKey)
-	if err != nil {
-		log.Infof("CheckAndRecover: Analysis: %+v, InstanceKey: %+v, candidateInstanceKey: %+v, "+
-			"skipProcesses: %v: NOT detecting/recovering host, could not obtain shard lock (%v)",
-			analysisEntry.Analysis, analysisEntry.AnalyzedInstanceKey, candidateInstanceKey, skipProcesses, err)
-		return false, nil, err
-	}
-	defer unlock(&err)
-
-	// Check if someone else fixed the problem.
-	tablet, err := TabletRefresh(analysisEntry.AnalyzedInstanceKey)
-	if err == nil && tablet.Type != topodatapb.TabletType_PRIMARY {
-		// TODO(sougou); use a version that only refreshes the current shard.
-		RefreshTablets()
+	// this check is needed because sometimes DeadPrimary code path is forcefully spawned off from other recoveries like PrimaryHasPrimary.
+	// So we need to check that we only run an ERS if the instance that we analyzed was actually a primary! Otherwise, we would end up running an ERS
+	// even when the cluster is fine or the problem can be fixed via some other recovery
+	if tablet.Type != topodatapb.TabletType_PRIMARY {
+		RefreshTablets(true)
 		AuditTopologyRecovery(topologyRecovery, "another agent seems to have fixed the problem")
-		// TODO(sougou): see if we have to reset the cluster as healthy.
 		return false, topologyRecovery, nil
 	}
 
-	AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("will handle DeadPrimary event on %+v", analysisEntry.ClusterDetails.ClusterName))
-	recoverDeadPrimaryCounter.Inc(1)
-	recoveryAttempted, promotedReplica, lostReplicas, err := recoverDeadPrimary(topologyRecovery, candidateInstanceKey, skipProcesses)
-	if err != nil {
-		AuditTopologyRecovery(topologyRecovery, err.Error())
+	// check if we have received SIGTERM, if we have, we should not continue with the recovery
+	val := atomic.LoadInt32(&hasReceivedSIGTERM)
```

Comment:
When VtOrc receives a SIGTERM it waits until all the locks have been released. The count of help locks is kept in an atomic variable `shardsLockCounter`. So in VtOrc, we first check whether we have received SIGTERM or not, if we have then we do not want to proceed further. If we haven't and decide to go ahead with ERS, then we increment the `shardsLockCounter`. If we now receive a SIGTERM, the we won't exit until all the. locks have been released which will only happen after ERS is complete. We do the SIGTERM check over here, because we do not want to go ahead with a new promotion when we have already received a SIGTERM

Reply:
My main point is that this check doesn't seem to add that much value. The sigterm could be received right after this check is done. 

The robust solution in my view is to make sure that on shut-down, it waits for all in-flight failovers to finish.

Reply:
It does wait using the `shardsLockCounter`. And we must also make it so that we do not start another failover while the existing failovers are being fixed, so we need to check whether we have received SIGTERM or not. Otherwise, there could be a case (purely fictional, but possible) that you have 2 shards and the first one is running an ERS. We receive a SIGTERM and are waiting for it to finish. Before this ERS fnishes, we started a ERS on the other shard and now have to wait for it to complete and this cycle can go on and on. To prevent, situations like this, we need both the checks

Reply:
Not sure if conveying my point clearly. Unless I'm missing something, that fictional situation you are describing it is still possible with this check. If that's the case... why should we add this check?

Reply:
The situation I am describing will not happen now, since we also check whether we have received SIGTERM or not. On receiving SIGTERM, we set `hasReceivedSIGTERM` to 1. In this case, we read that value and will not start a new ERS promotion. 
So, in the scenario I described, once we receive a SIGTERM, because of the `hasReceivedSIGTERM` check no more ERSs will be issued. We would just wait to release all the locks and finish the execution of the current ERSs.

Reply:
I agree that this check doesn't fix the problem; it can intercept a subset of scenarios where the signal is sent just before this specific step.

If I were to write `orchestrator` today, I'd more heavily rely on context `WithCancel`. This did not exist when `orchestrator` began, and I avoided a refactor when `context` became mainstream.